### PR TITLE
Global Styles: absorb editor settings transformation in WP_Theme_JSON

### PR DESCRIPTION
**Pull request by oandregal**
Extracted from https://github.com/WordPress/gutenberg/pull/30541 so it can land in pieces and make reviews easier/quicker.

This code deals with a theme.json structure and needs to know its internal shape. It's better this is part of the utils that deal with a theme.json array and not leaked to consumers. Further refactorings need to be done to offer higher-level APIs to consumers, but this is a first step in isolating model changes to a single point (WP_Theme_JSON).

## How to test

`npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php`

### CODE REVIEW

File: lib/class-wp-theme-json-resolver.php [hunk 1]
```diff
@@ -242,7 +242,6 @@ public static function get_core_data() {
 			return self::$core;
 		}
 
-		$all_blocks = WP_Theme_JSON::ALL_BLOCKS_NAME;
```
Comment (oandregal) [inline lib/class-wp-theme-json-resolver.php:245]: This was a leftover I found, so I thought I'd remove it here as well.

File: lib/global-styles.php [hunk 2]
```diff
@@ -5,107 +5,6 @@
  * @package gutenberg
  */
 
-/**
```
Comment (oandregal) [inline lib/global-styles.php:8]: This has been moved verbatim to `WP_Theme_JSON::get_from_editor_settings`, except for the first 4 lines:

```php
$all_blocks                                = WP_Theme_JSON::ALL_BLOCKS_NAME;
$theme_settings                            = array();
$theme_settings['settings']                = array();
$theme_settings['settings'][ $all_blocks ] = array();
```

that were collapsed into one:

```php
$theme_settings = array( 'settings' => array() );
```

File: phpunit/class-wp-theme-json-legacy-settings-test.php [hunk 3]
```diff
@@ -1,202 +0,0 @@
-<?php
-
-/**
- * Test that legacy settings are properly
- * reorganized into the theme.json structure.
- *
- * @package Gutenberg
- */
-
-class Theme_JSON_Legacy_Settings_Test extends WP_UnitTestCase {
```
Comment (oandregal) [inline phpunit/class-wp-theme-json-legacy-settings-test.php:10]: Tests were moved to `class-wp-theme-json-test.php` verbatim.

**Review by youknowriad — APPROVED** [review]
I personally don't like classic with a lot of static functions, why not just regular functions :) but I guess it's a way to gather related logic. 

I'd prefer more obvious patterns: factories, loaders, models... 

that said this PR doesn't bother me much, I think ultimately, this logic should be hidden entirely inside a "ThemeSettingsLoader" or something like that and the consumer won't need to know if the settings come from theme.json, user, or theme support flags, everything is merged and processed for him.

**Comment by oandregal** [comment]
I've been thinking about what would be the trade-offs of transforming `WP_Theme_JSON->get_stylesheet` into to `theme_json_to_stylesheet( $theme_json )` or something along those lines. Same for other functions such as `theme_json_merge`, `theme_json_from_editor_settings`, etc. Essentially, the opposite of what we have now: instead of having a single centralized point for low-level utilities, we have a set of independent low-level functions that operate on the last version of the theme.json structure, similar to how the PHP language has `array_merge`, `array_keys`, etc. Trade-offs that I see:

- Pollution of the global namespace. Besides having `theme_json_to_stylesheet` in the scope, we'll need to have some of the utility functions it uses to operate with individual nodes (which are now private methods of the class, hence don't pollute the global scope). Not sure what options do we have for this other than grouping them with classes and make the utility methods private (namespaces don't seem to have gained acceptance/popularity in WordPress core). Is it that strong a point as I think it is?

- Cross-cutting concerns. Some things are common to all of these methods such as schema migration or sanitization. It seems overkill to run these in every method call. Perhaps this could be solvable if the `theme_json_sanitization` and `theme_json_migrate_to_latest` methods attach some sort of information to the processed input, so the next time they see it, it can be skipped.

- Introducing changes to the schema. In addition to what we've talked at https://github.com/WordPress/gutenberg/pull/30600#discussion_r609448940 (introducing hook for every schema migrator) we'll have to introduce a hook for each one of these functions. If we can have a single point whose implementation can be entirely swapped with one hook it sounds a lot easier.

I agree there are other areas we need to improve applying what we've learned (higher-level APIs for consumers). I'm curious about your thoughts on these trade-offs for low-level APIs.

**Comment by youknowriad** [comment]
> Cross-cutting concerns. Some things are common to all of these methods such as schema migration or sanitization. It seems overkill to run these in every method call. Perhaps this could be solvable if the theme_json_sanitization and theme_json_migrate_to_latest methods attach some sort of information to the processed input, so the next time they see it, it can be skipped.

I actually think that's a bad pattern, it leads to dependencies on things that might have changed over time. 

> Introducing changes to the schema. In addition to what we've talked at #30600 (comment) (introducing hook for every schema migrator) we'll have to introduce a hook for each one of these functions. If we can have a single point whose implementation can be entirely swapped with one hook it sounds a lot easier.

I'm not sure I understand how this is related 😬 

--
I agree with the first point though in my mind, it's not a 1-1 translation between the functions we have now into new functions on the global scope. In my mind all of these are probably just private methods somewhere no one has access to, the consumer only gets access to a "loader" to get a config model and some public functions on the model (not static functions).

**Pull request merged**